modules/calib3d/src/stereosgbm.cpp

```diff
@@ -829,6 +822,645 @@ static void computeDisparitySGBM( const Mat& img1, const Mat& img2,
+                       PixType*& tmpBuf, CostType*& horPassCostVolume,
+                       CostType*& vertPassCostVolume, CostType*& vertPassMin, CostType*& rightPassBuf,
+                       CostType*& disp2CostBuf, short*& disp2Buf)
+{
+    // allocating all the required memory:
+    int costVolumeLineSize = width1*D;
+    int width1_ext = width1+2;
+    int costVolumeLineSize_ext = width1_ext*D;
+    int hsumBufNRows = SH2*2 + 2;
+
+    // main buffer to store matching costs for the current line:
+    int curCostVolumeLineSize = costVolumeLineSize*sizeof(CostType);
+
+    // auxiliary buffers for the raw matching cost computation:
+    int hsumBufSize  = costVolumeLineSize*hsumBufNRows*sizeof(CostType);
+    int pixDiffSize  = costVolumeLineSize*sizeof(CostType);
+    int tmpBufSize   = width*16*num_ch*sizeof(PixType);
+
+    // auxiliary buffers for the matching cost aggregation:
+    int horPassCostVolumeSize  = costVolumeLineSize_ext*sizeof(CostType); // buffer for the 2-pass horizontal cost aggregation
+    int vertPassCostVolumeSize = costVolumeLineSize_ext*sizeof(CostType); // buffer for the vertical cost aggregation
+    int vertPassMinSize        = width1_ext*sizeof(CostType);             // buffer for storing minimum costs from the previous line
+    int rightPassBufSize       = D*sizeof(CostType);                      // additional small buffer for the right-to-left pass
+
+    // buffers for the pseudo-LRC check:
+    int disp2CostBufSize = width*sizeof(CostType);
+    int disp2BufSize     = width*sizeof(short);
+
+    // sum up the sizes of all the buffers:
+    size_t totalBufSize = curCostVolumeLineSize +
+                          hsumBufSize +
+                          pixDiffSize +
+                          tmpBufSize  +
+                          horPassCostVolumeSize +
+                          vertPassCostVolumeSize +
+                          vertPassMinSize +
+                          rightPassBufSize +
+                          disp2CostBufSize +
+                          disp2BufSize +
+                          16;  //to compensate for the alignPtr shifts
+
+    if( buffer.empty() || !buffer.isContinuous() || buffer.cols*buffer.rows*buffer.elemSize() < totalBufSize )
+        buffer.create(1, (int)totalBufSize, CV_8U);
+
+    // set up all the pointers:
+    curCostVolumeLine  = (CostType*)alignPtr(buffer.ptr(), 16);
+    hsumBuf            = curCostVolumeLine + costVolumeLineSize;
+    pixDiff            = hsumBuf + costVolumeLineSize*hsumBufNRows;
+    tmpBuf             = (PixType*)(pixDiff + costVolumeLineSize);
+    horPassCostVolume  = (CostType*)(tmpBuf + width*16*num_ch);
+    vertPassCostVolume = horPassCostVolume + costVolumeLineSize_ext;
+    rightPassBuf       = vertPassCostVolume + costVolumeLineSize_ext;
+    vertPassMin        = rightPassBuf + D;
+    disp2CostBuf       = vertPassMin + width1_ext;
+    disp2Buf           = disp2CostBuf + width;
```

Comment:
@mshabunin I tried replacing this part with separate 16-byte-aligned memory allocations for each buffer. It still worked fine for me, i.e. no invalid memory accesses.